Common/MemoryUtil.cpp

```diff
@@ -234,52 +260,82 @@ void* AllocateAlignedMemory(size_t size,size_t alignment)
 	return ptr;
 }
 
-void FreeMemoryPages(void *ptr, size_t size)
-{
+void FreeMemoryPages(void *ptr, size_t size) {
+	if (!ptr)
+		return;
 	size = (size + 4095) & (~4095);
-	if (ptr)
-	{
 #ifdef _WIN32
-		if (!VirtualFree(ptr, 0, MEM_RELEASE))
-			PanicAlert("FreeMemoryPages failed!\n%s", GetLastErrorMsg());
+	if (!VirtualFree(ptr, 0, MEM_RELEASE))
+		PanicAlert("FreeMemoryPages failed!\n%s", GetLastErrorMsg());
 #elif defined(__SYMBIAN32__)
-		free(ptr);
+	free(ptr);
 #else
-		munmap(ptr, size);
+	munmap(ptr, size);
 #endif
-	}
 }
 
-void FreeAlignedMemory(void* ptr)
-{
-	if (ptr)
-	{
+void FreeAlignedMemory(void* ptr) {
+	if (!ptr)
+		return;
 #ifdef _WIN32
-		_aligned_free(ptr);
+	_aligned_free(ptr);
 #else
-		free(ptr);
+	free(ptr);
 #endif
-	}
 }
 
-void WriteProtectMemory(void* ptr, size_t size, bool allowExecute)
-{
-#ifdef _WIN32
-	DWORD oldValue;
-	if (!VirtualProtect(ptr, size, allowExecute ? PAGE_EXECUTE_READ : PAGE_READONLY, &oldValue))
-		PanicAlert("WriteProtectMemory failed!\n%s", GetLastErrorMsg());
+bool PlatformIsWXExclusive() {
+	// Only 64-bit iOS9 really needs this mode, but that's most iOS devices and all future ones,
+	// so let's keep things the same for all of them. Even without block linking, still should be much
+	// faster than IR JIT.
+#ifdef IOS
+	return true;
 #else
-	mprotect(ptr, size, allowExecute ? (PROT_READ | PROT_EXEC) : PROT_READ);
+	// Returning true here lets you test the W^X path on Windows and other non-W^X platforms.
+	return false;
 #endif
 }
 
-void UnWriteProtectMemory(void* ptr, size_t size, bool allowExecute)
-{
+void ProtectMemoryPages(const void* ptr, size_t size, uint32_t memProtFlags) {
+	INFO_LOG(JIT, "ProtectMemoryPages: %p (%d) : r%d w%d x%d", ptr, (int)size, (memProtFlags & MEM_PROT_READ) != 0, (memProtFlags & MEM_PROT_WRITE) != 0, (memProtFlags & MEM_PROT_EXEC) != 0);
+
+	if (PlatformIsWXExclusive()) {
+		if ((memProtFlags & (MEM_PROT_WRITE | MEM_PROT_EXEC)) == (MEM_PROT_WRITE | MEM_PROT_EXEC))
+			PanicAlert("Bad memory protect : W^X is in effect, can't both write and exec");
+	}
+	// Note - both VirtualProtect and mprotect will affect the full pages containing the requested range,
+	// so no need to round ptr and size down/up.
```

Comment:
First I thought it was, but turned out it wasn't, or at least it worked after I added the rounding...